ocaml/lambda/lambda.ml

```diff
@@ -1458,12 +1460,13 @@ let primitive_may_allocate : primitive -> alloc_mode option = function
   | Psetufloatfield _ -> None
   | Pduprecord _ -> Some alloc_heap
   | Pmake_unboxed_product _ | Punboxed_product_field _ -> None
+  | Pccall p when not Config.stack_allocation ->
+    if p.prim_alloc then Some alloc_heap else None
   | Pccall p ->
-     if not p.prim_alloc then None
-     else begin match p.prim_native_repr_res with
-       | (Prim_local|Prim_poly), _ -> Some alloc_local
-       | Prim_global, _ -> Some alloc_heap
-     end
+    begin match p.prim_native_repr_res with
+    | (Prim_local | Prim_poly), _ -> Some alloc_local
+    | Prim_global, _ -> Some alloc_heap
```

Comment:
This prevents the optimisation of non-local-returning noalloc C calls, I think there should be a noalloc check inside this line.

Reply:
Fixed (code now shared with Lambda)